mesonbuild/interpreter.py

```diff
@@ -2512,7 +2513,16 @@ def add_build_def_file(self, f):
         elif os.path.isfile(f) and not f.startswith('/dev'):
             srcdir = Path(self.environment.get_source_dir())
             builddir = Path(self.environment.get_build_dir())
-            f = Path(f).resolve()
+            try:
+                f = Path(f).resolve()
+            except OSError as e:
+                if e.errno == errno.EINVAL:
+                    # Magic WindowsApps executables aren't resolvable,
+                    # but still work... try making the path absolute
+                    # and hope for the best.
+                    f = Path(f).absolute()
```

Comment:
We need to find a way to detect whether the python wrapper in WindowsApps is an actual python or will just open a store link. Otherwise, https://github.com/mesonbuild/meson/pull/6459 will happen again. One way might be to check if the path is zero-sized or not:

```python
>>> pathlib.Path('C:/Users/nirbheek/AppData/Local/Microsoft/WindowsApps/python3.exe')
>>> s = p.stat()
>>> s
os.stat_result(st_mode=33279, st_ino=1970324836975681, st_dev=708247912, st_nlink=1, st_uid=0, st_gid=0, st_size=0, st_atime=1589576759, st_mtime=1589576759, st_ctime=1580153409)
>>> s.st_size
0
```

Reply:
To clarify, the same also happens with `python.exe`. On my machine this is what it shows when I don't have the store app installed. Please check what happens on your machine :)

Reply:
If reordering the search path and overlooking the path.resolve() error isn't acceptable, maybe an alternative approach would be to special-case looking for python or python3. In that case we could drop the WindowsApps path from the search as now, but then also inject dirname(sys.executable) at the end of the search paths.

Reply:
> In that case we could drop the WindowsApps path from the search as now, but then also inject dirname(sys.executable) at the end of the search paths.

I like this idea.

Reply:
Poke :)